password/src/Data/Password/Validate.hs

```diff
@@ -0,0 +1,237 @@
+-- >>> isValidPassword defaultPasswordPolicy defaultCharSetPredicate pass
+-- True
+isValidPassword :: PasswordPolicy -> CharSetPredicate -> Password -> Bool
+isValidPassword policy pre pass = null $ validatePassword policy pre pass
+{-# INLINE isValidPassword #-}
+
+-- | Check if given 'Password' fulfills all of the Policies, returns list of
+-- reasons why it's invalid.
+--
+-- >>> let pass = mkPassword "This_Is_Valid_Password1234"
+-- >>> validatePassword defaultPasswordPolicy defaultCharSetPredicate　pass
+-- []
+validatePassword :: PasswordPolicy -> CharSetPredicate -> Password -> [InvalidReason]
+validatePassword passwordPolicy@PasswordPolicy{..} charSetPredicate (Password password) =
+  catMaybes
+    [ isValidPolicy,
+      isTooShort,
+      isTooLong,
+      isUsingPolicyCharSetPredicate,
+      hasRequiredChar uppercaseChars Uppercase,
+      hasRequiredChar lowercaseChars Lowercase,
+      hasRequiredChar specialChars Special,
+      hasRequiredChar digitChars Digit
+    ]
+  where
+    isValidPolicy :: Maybe InvalidReason
+    isValidPolicy =
+      if isValidPasswordPolicy passwordPolicy charSetPredicate
+        then Nothing
+        else Just $ InvalidPasswordPolicy passwordPolicy
+    isTooLong :: Maybe InvalidReason
+    isTooLong =
+      if T.length password <= maximumLength
+        then Nothing
+        else Just $ PasswordTooLong maximumLength (T.length password)
+    isTooShort :: Maybe InvalidReason
+    isTooShort =
+      if T.length password >= minimumLength
+          then Nothing
+          else Just $ PasswordTooShort minimumLength (T.length password)
+    isUsingPolicyCharSetPredicate :: Maybe InvalidReason
+    isUsingPolicyCharSetPredicate =
+        let filteredText = T.filter (\c -> not $ (getCharSetPredicate charSetPredicate) c) password
+        in if T.null filteredText
+          then Nothing
+          else Just $ InvalidChar filteredText
+    hasRequiredChar :: Maybe Int -> CharacterCategory -> Maybe InvalidReason
+    hasRequiredChar Nothing _ = Nothing
+    hasRequiredChar (Just requiredCharNum) characterCategory =
+      let predicate = case characterCategory of
+            Uppercase -> isAsciiUpper
+            Lowercase -> isAsciiLower
+            Special   -> isSpecial
+            Digit     -> isDigit
+          actualRequiredCharNum = T.length $ T.filter predicate password
+       in if actualRequiredCharNum >= requiredCharNum
+          then Nothing
+          else Just $ NotEnoughReqChars characterCategory requiredCharNum actualRequiredCharNum
+
+-- | Checks if given 'PasswordPolicy' is valid
+--
+-- >>> isValidPasswordPolicy defaultPasswordPolicy defaultCharSetPredicate
+-- True
+isValidPasswordPolicy :: PasswordPolicy -> CharSetPredicate -> Bool
+isValidPasswordPolicy PasswordPolicy{..} charSetPredicate =
+  and
+   [ max minimumLength sumRequiredChars <= maximumLength
+   , minimumLength > 0
+   , maximumLength > 0
+   , isPositive uppercaseChars
```

Comment:
`all isPositive [lowercaseChars, uppercaseChars, specialChars, digitChars]`

Reply:
Though, to be honest, if someone sets an amount `< 0`, that would technically mean the same as not setting it.
If you want **at least** 0 characters, then it's fine with any amount of characters. Same with negative amounts.

And the `minimumLength` with 0 or less would also mean the same as just 0. Which should also be fine?

I'm very much of the opinion that the user should have all the flexibility they want, but give safe defaults and good warnings in the documentation.

Question: Would it be more advantageous to have passwords fail to validate at runtime, because we're being strict on the numbers having to be positive in the policy? Or have passwords be accepted, when the `minimumLength` or `specialChars` might be set to `-10`?

I can imagine scenarios where the policy could be decided programmatically, and if any of the "required character types" is `<= 0`, it would just mean "it's fine if there are none". And then having passwords fail to validate seems excessively strict. Especially since we're only returning `InvalidPasswordPolicy pwPolicy`, the user would somehow have to know that negative numbers or 0 is not allowed?
I'd rather allow those and have the documentation say that any number below 0 is treated as 0.

Then the only two things to check in a password policy is that the `maximumLength` is `> 0` and that it's more than `minimumLength`.